sblib/src/mem_mapper.cpp

```diff
@@ -0,0 +1,205 @@
+/*
+ * mem_mapper.cpp
+ *
+ *  Created on: Aug 16, 2015
+ *      Author: Deti Fliegl <deti@fliegl.de>
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License version 3 as
+ *  published by the Free Software Foundation.
+ *
+ */
+
+#include <sblib/platform.h>
+#include <sblib/internal/iap.h>
+#include <sblib/utils.h>
+#include <sblib/mem_mapper.h>
+#include <string.h>
+
+MemMapper::MemMapper(byte *flashBase = (byte*) 0xf000, unsigned int flashSize =
+        4096) :
+        flashBase(flashBase), flashSize(flashSize)
+{
+    flashSizePages = flashSize / FLASH_PAGE_SIZE;
+    flashBasePage = ((unsigned int) flashBase) / FLASH_PAGE_SIZE;
+    lastAllocated = 0; // means: nothing allocated in this run
+    writePage = 0;
+
+    memcpy(allocTable, flashBase, FLASH_PAGE_SIZE);
+}
+
+int MemMapper::doFlash(void)
+{
+    int ret = 0;
+    if (allocTableModified)
+    {
+        if (iapErasePage(flashBasePage) != IAP_SUCCESS)
+        {
+            fatalError();
+        }
+        if (iapProgram(flashBase, allocTable, FLASH_PAGE_SIZE) != IAP_SUCCESS)
+        {
+            fatalError();
+        }
+        allocTableModified = false;
+        ret |= 1;
+    }
+    if (flashMemModified)
+    {
+        if (iapErasePage(writePage) != IAP_SUCCESS)
+        {
+            fatalError();
+        }
+        if (iapProgram((byte *) (writePage << 8), writeBuf, FLASH_PAGE_SIZE)
+                != IAP_SUCCESS)
+        {
+            fatalError();
+        }
+        flashMemModified = false;
+        ret |= 2;
+    }
+    return ret;
+}
+
+bool MemMapper::isValidAddress(int virtAddress)
```

Comment:
On 19/08/15 07:51, Stefan Taferner wrote:
> I think you could rename it to isValidPage() as you always pass virtPage
> instead of virtAddress (see below).
> You could then make it an inline method as it comes down to "return
> virtPage<FLASH_PAGE_SIZE".
Removed this method now completely as it is just a simple comparison.